### ISSUE TITLE
Support query variant

### ISSUE THREAD

**Issue by Virgiel** [cornucopia-rs/cornucopia#164]
It is common to have SQL queries with different filters, currently I write each variant by hand:

```sql
--! raw_page_fitness: ProRawRow
select ... from pro ORDER BY fitness ASC LIMIT 51 OFFSET :offset;
--! raw_page_created: ProRawRow
select ... from pro ORDER BY created DESC LIMIT 51 OFFSET :offset;
--! raw_page_name: ProRawRow
select ... from pro ORDER BY lname ASC LIMIT 51 OFFSET :offset;
```

I see two ways to handle these cases:

# Runtime query

We should allow the user to replace the statement query with their own runtime generated queries, this should be advertised as a very advanced feature as it will cause runtime panic if the columns and parameters do not match.

```rust
pub struct Stmt {
    query: &'static str,
    cached: Option<Statement>,
}

```

Would become:

```rust
pub struct Stmt<'a> {
    query: &'a str,
    cached: Option<Statement>,
}

impl <'a> Stmt<'a> {
    pub fn runtime_query(&mut self, query: &'a) {
        self.query = query;
    }
}

```

# Pattern query

The user declares a pattern query and all possible variants that can complete the pattern. This way we check all variants at compile time! The runtime query solution could still be proposed for case we do not support.

```sql
--@ RawOrder
--: None 
--: Fitness : ORDER BY fitness ASC
--: Created : ORDER BY created DESC
--: Name : ORDER BY lname ASC

--! raw_page: ProRawRow
select ... from pro @RawOrder LIMIT 51 OFFSET :offset;
```

Would generate :

```rs
pub fn raw_page(raw_order: RawOrder) -> RawPageStmt {
    let query = match raw_order {...};
    RawPageStmt(cornucopia_async::private::Stmt::new(query))
}
```

**Comment by LouisGariepy** [comment]
Oh btw, do you want to add this to the 0.9 milestone, or is this for "the future"?

**Comment by ricardodarocha** [comment]
I agree with @LouisGariepy That increases complexibility. I suggest a layer to you generate sql files from templates. For instance, using jinja etc.

First you define templates and rules for that templates
Then you make a cli to process the template, generating the sql you want

**template**
`select * from pro ORDER BY {{ FILTER }}  ASC LIMIT 51 OFFSET :offset;`

**args.toml**
`FILTER: [fitness, created, lname]`

**your custom cli**
`$ generate_sql template.txt  args.toml -d \sql

You can easily create a cli like this with clap

**Comment by jacobsvante** [comment]
> # Pattern query
> 
> The user declares a pattern query and all possible variants that can complete the pattern. This way we check all variants at compile time! The runtime query solution could still be proposed for case we do not support.
> 
> ```sql
> --@ RawOrder
> --: None 
> --: Fitness : ORDER BY fitness ASC
> --: Created : ORDER BY created DESC
> --: Name : ORDER BY lname ASC
> 
> --! raw_page: ProRawRow
> select ... from pro @RawOrder LIMIT 51 OFFSET :offset;
> ```

This seems like a very good start. Builds on cornucopia's main strength - code generation time query verification! (I would call what sqlx does "compile time query verification", which wastes a lot more time on CPU / my time).

**Comment by LouisGariepy** [comment]
Cross-reference: similar feature requested upstream in the merged Clorinde fork at [halcyonnouveau/clorinde#63](https://github.com/halcyonnouveau/clorinde/issues/63) — the `[[ ... ]]` templating syntax for optional `WHERE` / `SET` fragments (inspired by Metabase). That thread has additional discussion worth folding in here:

- A workaround using `(:type::TEXT IS NULL OR type = :type)` that compiles correctly.
- A suggestion to go further with full Jinja/dbt-style templating (covering halcyonnouveau/clorinde#47).
- A `WHERE true [[ AND ... ]]` / `SET id = id [[ , ... ]]` trick (from sqlc) to keep partial templates syntactically valid when all variables are `None`.

Not migrating it as a separate issue since it overlaps with the pattern-query approach already discussed here, but flagging it so the prior art isn't lost.
